Patatrack integration - calorimeters shared code (6/N)

fwyzard opened the pull request:
#### PR description:

Data formats and algorithms shared by the ECAL and HCAL local reconstruction (multifit and MAHI).

#### PR validation:

Changes in use in the Patatrack releases.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Includes changes from:
  - cms-patatrack#252: Skip CUDA-related tests if no GPU is present
  - cms-patatrack#337: Implement inversion by Cholesky Decomposition
  - cms-patatrack#408: Replace cuda::device operations with native CUDA calls
  - cms-patatrack#423: Rename exitSansCUDADevices to requireCUDADevices
  - cms-patatrack#429: Implement changes from the CUDA framework review
  - cms-patatrack#442: Integrate the comments from the upstream PRs
  - cms-patatrack#502: Update ECAL and HCAL reconstruction to run on multple GPUs [1/3]
  - cms-patatrack#523: Refactor common ECAL and HCAL code
  - cms-patatrack#557: Move multifit/MAHI common code to DataFormats/CaloRecHit

fwyzard committed:
Synchronise with CMSSW_10_4_0_pre2

fwyzard committed:
Skip CUDA-related tests if no GPU is present (#252)

Make unit tests that require a CUDA device skip the test and exit
succesfully if the CUDA runtime is not available, or no CUDA devices
are available.

fwyzard committed:
Apply clang-format style formatting

fwyzard committed:
Rename exitSansCUDADevices to requireCUDADevices (#423)

makortel committed:
Implement changes from the CUDA framework review (#429)

Rename the cudautils namespace to cms::cuda or cms::cudatest, and drop the CUDA prefix from the symbols defined there.

Always record and query the CUDA event, to minimize need for error checking in CUDAScopedContextProduce destructor.

Add comments to highlight the pieces in CachingDeviceAllocator that have been changed wrt. cub.

Various other updates and clean up:
  - enable CUDA for compute capability 3.5.
  - clean up CUDAService, CUDA tests and plugins.
  - add CUDA existence protections to BuildFiles.
  - mark thread-safe static variables with CMS_THREAD_SAFE.

fwyzard committed:
Synchronise with CMSSW_11_1_0_pre7

vkhristenko committed:
Update ECAL and HCAL reconstruction to run on multple GPUs [1/3] (#502)

Use caching allocators for host and device CUDA memory.
Use dedicated ESProducers to make part of the modules' configuration available on all GPUs.
Rename hcal and hcal::common namespaces to to calo::common.

fwyzard committed:
Apply code formatting

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
For all questions, please address @amassiro @mariadalfonso @vkhristenko .
For all changes, please make PRs against `cms-patatrack:patatrack_integration_6_N_calorimeters_common` .

fwyzard commented:
@cmsbuild, please test with #31703

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31704/18866

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CUDADataFormats/CaloCommon/BuildFile.xml  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/interface/Common.h  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/src/classes.h  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/src/classes_def.xml  modified in PR(s): #27983
  - File  DataFormats/Math/interface/EigenComputations.h  modified in PR(s): #27983
  - File  DataFormats/Math/test/BuildFile.xml  modified in PR(s): #27983

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-sw/cmssw#31703

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

CUDADataFormats/CaloCommon
DataFormats/Math


The following packages do not have a category, yet:

CUDADataFormats/CaloCommon
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@perrotta, @jpata, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@makortel, @felicepantaleo, @fabiocos, @rovere this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

Tested at: f6571a8f917c2cc84f96b2762088eb291930d3c2

CMSSW: CMSSW_11_2_X_2020-10-07-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-aa7054/9784/summary.html

I found follow errors while testing this PR

**Failed tests:  HeaderConsistency**

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-aa7054/9784/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2542225
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2542202
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
@cmsbuild, please test with #31703

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31704/18873

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CUDADataFormats/CaloCommon/BuildFile.xml  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/interface/Common.h  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/src/classes.h  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/src/classes_def.xml  modified in PR(s): #27983
  - File  DataFormats/CaloRecHit/test/BuildFile.xml  modified in PR(s): #27983
  - File  DataFormats/CaloRecHit/test/test_calo_rechit.cu  modified in PR(s): #27983
  - File  DataFormats/Math/interface/EigenComputations.h  modified in PR(s): #27983
  - File  DataFormats/Math/test/BuildFile.xml  modified in PR(s): #27983

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-sw/cmssw#31703

cmsbuild commented:
Pull request #31704 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

cmsbuild commented:
-1

Tested at: d07d03072ee00e32717cd2f05b0ce6b76da1bec2

CMSSW: CMSSW_11_2_X_2020-10-07-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-aa7054/9793/summary.html

I found follow errors while testing this PR

**Failed tests:  HeaderConsistency**

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-aa7054/9793/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2542225
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2542190
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jpata commented on DataFormats/CaloRecHit/test/test_calo_rechit.cu:
```diff
@@ -1,10 +1,11 @@
+#include <cassert>
+#include <iostream>
```
not needed?

fwyzard commented:
The new packages for this and the other Patatrack PRs should be taken care of by cms-sw/cms-bot#1395 .

jpata commented on DataFormats/Math/interface/EigenComputations.h:
```diff
@@ -0,0 +1,440 @@
+#include <Eigen/Dense>
+
+namespace calo {
+  namespace multifit {
```
this namespace name seems to specific for the header

fwyzard replied:
Possibly.
Is that a problem ?

jpata replied:
It seems odd to put generic linear algebra functions in `calo::multifit`. From the top of my head, why not something like `reco::eigen`? Or otherwise, if these are really calorimeter-specific functions with no use elsewhere, then this header should be renamed/moved to denote that.
Let me know if this will be addressed.

fwyzard replied:
moved to a calorimeter-specific package

jpata commented on DataFormats/Math/interface/EigenComputations.h:
```diff
@@ -0,0 +1,440 @@
+
+      __forceinline__ __device__ MapSymM(T* data) : data{data} {}
+
+      __forceinline__ __device__ T const& operator()(int const row, int const col) const {
```
it would be nice to have unit tests the functions defined in this header

jpata commented on DataFormats/Math/interface/EigenComputations.h:
```diff
@@ -0,0 +1,440 @@
+    }
+
+    /*
+    // TODO: add active bxs
```
please remove commented out code, or add a comment why this is needed and when this will be implemented

mariadalfonso replied:
Similar common call  done in 

* ECAL https://github.com/cms-sw/cmssw/pull/31719/files#diff-98f76d25429b5059efd61f6632594e42R17
RecoLocalCalo/EcalRecProducers/plugins/KernelHelpers.h

* HCAL https://github.com/cms-sw/cmssw/pull/31720/files#diff-8ad73af139b7dd05aaecb810d8327c62R625
RecoLocalCalo/HcalRecProducers/src/MahiGPU.cu

would be better here in a common place in DataFormats/Math/interface/EigenComputations.h

the only differences ( originally from the CPU code) are in the 
* nan check (only on ECAL, but would benefit in both)
* convergence parameters (those two integer can be configurable)

       ++iter;
         if (iter % 16 == 0)
           eps_to_use *= 2;

      if (iter % 10 == 0)
           eps_to_use *= 10;

jpata replied:
Thanks for pointing out the code duplications across the PRs. I was following the links a bit, and I would like to ask for a single, generic implementation of these functions:

- https://github.com/cms-sw/cmssw/pull/31704/files/d07d03072ee00e32717cd2f05b0ce6b76da1bec2#diff-f456782d8083d5d31604451980616cc7728f6be06f43f371e0ef1c67eea2dfdcR282
- https://github.com/cms-sw/cmssw/pull/31720/files#diff-4c3f70988bc0587761bb4c18f9a393f6ef483e3601452ee15e5cd3d34654d1f6R625
- https://github.com/cms-sw/cmssw/pull/31719/files#diff-ed9059ebaf6819819074dce32f5dca13fc1499b457a6a90031adac0b419bcd32R17

This commented-out code block above can then be dropped, and you can reduce the number of lines in the other PRs. Let me know if you will address this, or did I misunderstand the comment?

mariadalfonso replied:
ok, generic implementation of the fnnls, where you want to place ?
not on ECAL, not on HCAL
is fine to have in DataFormats/Math/interface/EigenComputations.h together in the calo::multifit ?
or where you suggest to add all these common fits function ?

jpata replied:
If this is generic linear algebra, just keep the implementation in DataFormats/Math/interface/EigenComputations.h. I don't want to get stuck on the namespace name here, if there is no better idea in sight and it becomes a blocker for moving forward.

The higher priority item is to remove this code duplication (if possible and feasible) from the other PRs. Thank you for pointing out this particular example, are there other functions which are similarly duplicated? We can of course invest in some automatic checks, but it would be good to hear from the authors.

fwyzard replied:
duplicate code removed

jpata reviewed (commented):
Thanks for this PR. In addition to the comments inline, can you take care of the header consistency issue in the tests?

mariadalfonso commented:
The function
eigen_solve_submatrix(SampleMatrix& mat, SampleVector& invec, SampleVector& outvec, unsigned NP){...}
in RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationKernels.cu
https://github.com/cms-sw/cmssw/pull/31719/files#diff-605d8e9b7dfe62cdefc034c32e7e0e03R20

Would be better placed here in DataFormats/Math/interface/EigenComputations.h
This is the direct translation of HCAL and ECAL cpu code, that is currently not used on gpu since is replaced by calo::multifit::compute_decomposition_forwardsubst_with_offsets and calo::multifit::update_decomposition_forwardsubst_with_offsets

fwyzard commented:
@jpata:
  - do the comments from @mariadalfonso answer your questions ?
  - do you have further comments ?

From my side I would add the that SoA approach used here (and in the pixel code) will be reviewed and likely updated after the integration; together with @makortel we are evaluating how to better discuss the topic in an upcoming Core Software meeting.

jpata commented:
Please see inline for my replies, and let me know how you will proceed before closing the comment. In addition:
- please address integration test failures (header consistency)
- is a lack of unit tests in such DataFormats code acceptable, @slava77 et al?

Thanks!

slava77 commented:
> is a lack of unit tests in such DataFormats code acceptable, @slava77 et al?

I was never diligent about unit tests.
Having some would be nice, especially given that this is a DF/Math

jpata commented:
regarding unit tests: I want to be consistent with reviews of this PR wrt. other PRs in reco view. If there is no strict requirement on unit tests in DF/Math but it's rather a nice-to-have, then I don't want to put this as a requirement here today.
The other review comments stand.

jpata commented:
@amassiro @mariadalfonso @vkhristenko please let us know if you plan to implement the changes requested in the review? Just to make sure things don't get stuck. Thank you.

fwyzard commented:
Include guards and relative comments should be fixed by the latest commit.

jpata commented:
pinging Core: @makortel @smuzaffar 
Could you help us with the header consistency issue here: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-aa7054/9784/headers_chks.log?

Something with cuda-specific patterns not being detected correctly. Thank you!

smuzaffar commented:
@jpata , the newly added file `DataFormats/Math/interface/EigenComputations.h` does not compile with gcc . This file is not currently included by any cc files that is why `scram build` works but if you include this in any cc file then it will fail.

So is this file only suppose to be included via a cuda source `.cu` file?

fwyzard commented:
> So is this file [`DataFormats/Math/interface/EigenComputations.h`] only suppose to be included via a cuda source `.cu` file?

Currently, yes.

Do we need to make all header files parsable by GCC ?

smuzaffar commented:
If it is in interface directory then this means it can be included by any one and then yes it needs to be parsable by GCC. If this is only included via cuda sources then may be we can change the file extension?

fwyzard commented:
OK.

@mariadalfonso has provided fixes for some of the comments via https://github.com/cms-patatrack/cmssw/pull/557 .

Once that is validated and added to #31704, #31719 and #31720 I will see how to make the header more friendly towards non-CUDA compilers.

mariadalfonso committed:
Move multifit/MAHI common code to DataFormats/CaloRecHit (#557)

Move multifit/MAHI common code to DataFormats/CaloRecHit/interface/MultifitComputations.h .
Improve naming and description of fnnls parameters.
Use Eigen preprocessor symbols instead of explicit CUDA keywords, and CUDA preprocessor symbols to protect CUDA-only functions.

Co-authored-by: Andrea Bocci <andrea.bocci@cern.ch>

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
@cmsbuild, please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31704/19234

- This PR adds an extra 20KB to repository

- Found files with invalid states:
  - DataFormats/Math/interface/EigenComputations.h:
    - Added: d07d03072ee00e32717cd2f05b0ce6b76da1bec2
    - Deleted: 02d87a4557706f5785566b0fb3747f9d4a11f872


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CUDADataFormats/CaloCommon/BuildFile.xml  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/interface/Common.h  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/src/classes.h  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/src/classes_def.xml  modified in PR(s): #27983
  - File  DataFormats/CaloRecHit/test/BuildFile.xml  modified in PR(s): #27983
  - File  DataFormats/CaloRecHit/test/test_calo_rechit.cu  modified in PR(s): #27983
  - File  DataFormats/Math/test/BuildFile.xml  modified in PR(s): #27983

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #31704 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 02d87a4557706f5785566b0fb3747f9d4a11f872
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bad05c/10117/summary.html
CMSSW: CMSSW_11_2_X_2020-10-19-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

fwyzard commented:
AFAICT all comments have been addressed, and all automatic tests pass

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bad05c/10117/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544100
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544077
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
Cleaned up the history to avoid adding and removing `DataFormats/Math/interface/EigenComputations.h` .

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31704/19245

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CUDADataFormats/CaloCommon/BuildFile.xml  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/interface/Common.h  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/src/classes.h  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/src/classes_def.xml  modified in PR(s): #27983
  - File  DataFormats/CaloRecHit/test/BuildFile.xml  modified in PR(s): #27983
  - File  DataFormats/CaloRecHit/test/test_calo_rechit.cu  modified in PR(s): #27983


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31704/19245/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31704/19245/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31704/19247

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CUDADataFormats/CaloCommon/BuildFile.xml  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/interface/Common.h  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/src/classes.h  modified in PR(s): #27983
  - File  CUDADataFormats/CaloCommon/src/classes_def.xml  modified in PR(s): #27983
  - File  DataFormats/CaloRecHit/test/BuildFile.xml  modified in PR(s): #27983
  - File  DataFormats/CaloRecHit/test/test_calo_rechit.cu  modified in PR(s): #27983

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #31704 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

jpata commented:
Thanks @smuzaffar for the help and @fwyzard for addressing the review items! Code checks pass - good. Let's see how the tests go, I will also try to run a manual test on my local GPU machine.

For my education about "please test" for patatrack, is there any standard (or extra) workflow where this new code is enabled? Should the test now be run with "enable gpu"?

fwyzard commented:
> For my education about "please test" for patatrack, is there any standard (or extra) workflow where this new code is enabled? Should the test now be run with "enable gpu"?

This PR adds only new data formats, so there are no workflows associated.

#31719 and #31720 should enable the .512 and .522 workflows to run on GPUs.

I'm not sure what is the list of workflows and tests that "enable gpu" adds.

cmsbuild commented:
+1
Tested at: 8938c570e40f46eb8a0a7ed74e7fbac07f0f183a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bad05c/10126/summary.html
CMSSW: CMSSW_11_2_X_2020-10-19-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bad05c/10126/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544100
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544065
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

cmsbuild commented:
Pull request #31704 was updated. @perrotta, @makortel, @jpata, @fwyzard, @slava77 can you please check and sign again.

fwyzard commented:
+1

jpata commented:
The tests pass, but there is this in the [log](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bad05c/10126/unitTests.log)
```
===== Test "test_calo_rechit" ====
Failed to initialise the CUDA runtime, the test will be skipped.
```
let's see if enabling GPU helps?

jpata commented:
@cmsbuld enable gpu

jpata commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **ENABLE_BOT_TESTS** = GPU

jpata commented:
I was able to run locally at least (had to install the latest 455.23.05 driver):

```
$ ./test/slc7_amd64_gcc820/test_calo_rechit
0, 10 GeV, 1 ns  flags=0x0  aux=0x0
0, 10 GeV, 1 ns  flags=0x0  aux=0x0
all good!
```

cmsbuild commented:
+1
Tested at: 8938c570e40f46eb8a0a7ed74e7fbac07f0f183a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bad05c/10144/summary.html
CMSSW: CMSSW_11_2_X_2020-10-19-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bad05c/10144/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544100
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544077
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jpata commented:
Even with enabling the GPU in the test, `test_calo_rechit` did not run. I think we need to raise this as a separate issue (unless I misunderstood how this is supposed to work), but it's not a blocker for this PR as I have been able to verify the test runs locally.

```
===== Test "test_calo_rechit" ====
Failed to initialise the CUDA runtime, the test will be skipped.

---> test test_calo_rechit succeeded
```

jpata commented:
+reconstruction

- introduces `DataFormats/CaloRecHit` which contains shared code for calorimeters for the patatrack integration
- unit tests are present for a part of the code and have been verified to run independently
- code checks and tests pass, so it won't break the build
- minor review items (namespace & file names, commented-out code for `fnnls`) have been addressed
- no reco differences are expected (this code does not run in any workflow) or observed
- this PR is urgent for CMSSW_11_2_X_pre_8 in the timescale of today

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
@jpata @fwyzard you might create an issue with the list of the "nice-to-have" features after the integration of each Patatrack PR (eg. `EigenComputations` unit tests)

cmsbuild commented:
+1
Tested at: 8938c570e40f46eb8a0a7ed74e7fbac07f0f183a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bad05c/10227/summary.html
CMSSW: CMSSW_11_2_X_2020-10-22-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison not run due to Build errors/Fireworks only changes/No short matrix requested (RelVals and Igprof tests were also skipped)

davidlange6 commented:
looks likely that the clang build will complain about this PR once its new headers are used - we see this in the modules IB

  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/9faf547acdcdd0e319dedacbbb9eebbe/opt/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_CXXMODULE_X_2020-10-26-2300/src/DataFormats/CaloRecHit/interface/MultifitComputations.h:341:58: warning: '&&' within '||' [-Wlogical-op-parentheses]
           if (w_max < eps || w_max_idx == w_max_idx_prev && w_max == w_max_prev)
                          ~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/9faf547acdcdd0e319dedacbbb9eebbe/opt/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_CXXMODULE_X_2020-10-26-2300/src/DataFormats/CaloRecHit/interface/MultifitComputations.h:341:58: note: place parentheses around the '&&' expression to silence this warning
          if (w_max < eps || w_max_idx == w_max_idx_prev && w_max == w_max_prev)
                                                         ^
                             (                                                 )
